File: reference/lib/cost-log/report.js
// Prints a summary of logged LLM costs. Run with: npm run cost:report

import { readFileSync, existsSync } from 'node:fs';
import { join } from 'node:path';

function logPath() {
  const root = process.env.AUTOMATION_ROOT || process.cwd();
  return join(root, 'logs', 'cost-log.jsonl');
}

function main() {
  const path = logPath();
  if (!existsSync(path)) {
    console.log('No cost log yet at', path);
    return;
  }
  const lines = readFileSync(path, 'utf8').split('\n').filter(Boolean);
  const runs = lines.map((l) => JSON.parse(l));

  const byMonth = new Map();
  let totalCost = 0;
  let totalTokens = 0;

  for (const r of runs) {
    const month = r.ts.slice(0, 7);
    const agg = byMonth.get(month) || { runs: 0, cost: 0, tokens: 0 };
    agg.runs++;
    agg.cost += r.costUsd || 0;
    agg.tokens += r.totalTokens || 0;
    byMonth.set(month, agg);
    totalCost += r.costUsd || 0;
    totalTokens += r.totalTokens || 0;
  }

  console.log('LLM cost report');
  console.log('================');
  console.log(`Runs: ${runs.length}   Tokens: ${totalTokens.toLocaleString()}   Cost: $${totalCost.toFixed(4)}`);
  console.log('');
  console.log('By month:');
  for (const [month, agg] of [...byMonth.entries()].sort()) {
    console.log(`  ${month}: ${agg.runs} runs, ${agg.tokens.toLocaleString()} tokens, $${agg.cost.toFixed(4)}`);
  }
  const lastReal = runs.filter((r) => !r.dryRun).at(-1);
  if (lastReal) {
    console.log('');
    console.log(`Last real run: ${lastReal.ts} (${lastReal.model}) $${lastReal.costUsd.toFixed(6)} for ${lastReal.itemCount} items`);
  }
}

main();

File: internal/obs/report.go
package obs

import (
	"bufio"
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"path/filepath"
	"sort"

	"github.com/sl6117/automations/internal/storage"
)

// costLogPath is the same file LogRun writes to
func costLogPath() string {
	return filepath.Join(logRoot(), "logs", "cost-log.jsonl")
}

// LoadRuns reads every logged run. A missing log isn't an error (returns nil)
func LoadRuns(ctx context.Context, store storage.Store) ([]Run, error) {
	data, err := store.Get(ctx, costLogKey)
	if errors.Is(err, storage.ErrNotExist) {
		return nil, nil
	}
	if err != nil {
		return nil, fmt.Errorf("read cost log: %w", err)
	}

	var runs []Run
	scanner := bufio.NewScanner(bytes.NewReader(data))

	for scanner.Scan() {
		line := bytes.TrimSpace(scanner.Bytes())
		if len(line) == 0 {
			continue
		}
		var r Run
		if err := json.Unmarshal(line, &r); err != nil {
			return nil, fmt.Errorf("unmarshal cost log: %w", err)
		}
		runs = append(runs, r)
	}
	return runs, nil
}

type monthAgg struct {
	runs   int
	tokens int
	reads  int
	cost   float64
}

// Report aggregates the cost log and writes a human-readable summary to w
func Report(ctx context.Context, store storage.Store, w io.Writer) error {
	runs, err := LoadRuns(ctx, store)
	if err != nil {
		return err
	}
	if len(runs) == 0 {
		fmt.Fprintf(w, "No cost log yet at %s\n", costLogPath())
		return nil
	}

	byMonth := map[string]*monthAgg{}
	var totalCost float64
	var totalReads int
	var totalTokens int
	var lastReal *Run

	for i := range runs {
		r := runs[i]
		tokens := r.InputTokens + r.OutputTokens
		month := r.Timestamp

		if len(month) >= 7 {
			month = month[:7]
		}
		agg := byMonth[month]
		if agg == nil {
			agg = &monthAgg{}
			byMonth[month] = agg
		}
		agg.runs++
		agg.tokens += tokens
		agg.reads += r.SourceReads
		agg.cost += r.CostUSD
		totalCost += r.CostUSD
		totalTokens += tokens
		totalReads += r.SourceReads

		if !r.DryRun {
			rr := r
			lastReal = &rr
		}
	}

	fmt.Fprintln(w, "LLM Cost Report")
	fmt.Fprintln(w, "================")
	fmt.Fprintf(w, "Runs: %d   Tokens: %d   X reads: %d   Cost: $%.4f\n\n", len(runs), totalTokens, totalReads, totalCost)
	fmt.Fprintln(w, "By month:")

	months := make([]string, 0, len(byMonth))
	for month := range byMonth {
		months = append(months, month)
	}
	sort.Strings(months)
	for _, m := range months {
		agg := byMonth[m]
		fmt.Fprintf(w, "  %s: %d runs, %d tokens, %d reads, $%.4f\n", m, agg.runs, agg.tokens, agg.reads, agg.cost)
	}

	if lastReal != nil {
		fmt.Fprintf(w, "\nLast real run: %s (%s) $%.6f for %d items\n",
			lastReal.Timestamp, lastReal.Model, lastReal.CostUSD, lastReal.ItemCount)
	}
	return nil

}
